homeassistant/components/ring/coordinator.py

```diff
@@ -0,0 +1,102 @@
+
+from .const import NOTIFICATIONS_SCAN_INTERVAL, SCAN_INTERVAL
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def _call_api(hass: HomeAssistant, target, *args, msg_suffix=""):
+    try:
+        return await hass.async_add_executor_job(target, *args)
+    except ring_doorbell.AuthenticationError as err:
+        # Raising ConfigEntryAuthFailed will cancel future updates
+        # and start a config flow with SOURCE_REAUTH (async_step_reauth)
+        raise ConfigEntryAuthFailed from err
+    except ring_doorbell.RingTimeout as err:
+        raise UpdateFailed(
+            f"Timeout communicating with API{msg_suffix}: {err}"
+        ) from err
+    except ring_doorbell.RingError as err:
+        raise UpdateFailed(f"Error communicating with API{msg_suffix}: {err}") from err
+
+
+@dataclass
+class RingDeviceData:
+    """RingDeviceData."""
+
+    device: RingGeneric
+    history: Optional[list] = None
+
+
+class RingDataCoordinator(DataUpdateCoordinator[dict[int, RingDeviceData]]):
+    """Base class for device coordinators."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        ring_api,
+    ) -> None:
+        """Initialize my coordinator."""
+        super().__init__(
+            hass,
+            name="devices",
+            logger=_LOGGER,
+            update_interval=SCAN_INTERVAL,
+        )
+        self.ring_api = ring_api
+        self.first_call = True
+
+    async def _async_update_data(self):
+        """Fetch data from API endpoint."""
+        update_method = "update_data" if self.first_call else "update_devices"
+        await _call_api(self.hass, getattr(self.ring_api, update_method))
+        self.first_call = False
+        data = {}
+        devices = self.ring_api.devices()
+        for device_type in devices:
+            for device in devices[device_type]:
+                # Don't update all devices in the ring api, only those that set
+                # their device id as context when they subscribed.
+                if device.id in set(self.async_contexts()):
+                    data[device.id] = RingDeviceData(device=device)
+                    if hasattr(device, "history"):
+                        data[device.id].history = await _call_api(
+                            self.hass,
+                            lambda device: device.history(limit=10),
+                            device,
+                            msg_suffix=f" for device {device.name}",  # device_id is the mac
+                        )
+                    await _call_api(
+                        self.hass,
+                        device.update_health_data,
+                        msg_suffix=f" for device {device.name}",
+                    )
+        return data
+
+
+class RingNotificationsCoordinator(DataUpdateCoordinator[None]):
+    """Global notifications coordinator."""
+
+    def __init__(self, hass: HomeAssistant, ring_api) -> None:
```

Comment:
```suggestion
    def __init__(self, hass: HomeAssistant, ring_api: ring_doorbell.Ring) -> None:
```

Would it be an idea to add typings to cade you touch as well? Not a requirement, but would make the code a lot easier to read. And perhaps eventually (in a follow-up PR) you can get rid of the broad `import ring_doorbell` as well.

Reply:
Done, I've added typing to all touched code.  Regarding the broad import of ring_doorbell I do that because when I then mock/patch for testing I've always patched in the right place, i.e patch("ring_doorbell.SomeRingObject") always works and I don't have to patch each module that imports a ring_doorbell object directly.  What do you think?

Reply:
Resolving conversation as typing added to all touched code.